clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -543,13 +566,68 @@ synchronized void updateLastAckedOffset(ProduceResponse.PartitionResponse respon
         }
     }
 
+    public synchronized void handleCompletedBatch(ProducerBatch batch, ProduceResponse.PartitionResponse response) {
```

Comment:
Just to clarify this function is just moved here, no logical changes?

Also if it is indeed the case, could you make some comments when creating the PR for ease of review :) ?

Reply:
Yes, it is moved to simplify the TransactionManager API so that I could write better test cases. Otherwise it was very difficult to hit the cases without effectively rewriting this logic in the test case.